<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
# Импорт библиотек
import pandas as pd
import psycopg2
import plotly
import plotly.express as px

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
SELECT count(v.id) as count_vacancies -- подсчет количества вакансий в таблице vacancies
  FROM vacancies as v
'''

In [5]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\4045908245.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


count_vacancies
0            49197

Выводы: количество вакансий в исследуемой выборке ~ 50 000

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
SELECT count(e.id) as count_employers -- подсчет количества работодателей в таблице employers
  FROM employers as e
'''

In [7]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\533540996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


count_employers
0            23501

Выводы: количество работодателей в исследуемой выборке ~ 24 000. Что в 2 раза меньше количества вакансий, отсюда можно сделать вывод, что большинству работодателей принадлежит 2 и более вакансии.

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
SELECT count(a.id) as count_areas -- подсчет количества регионов в таблице areas
  FROM areas as a
'''

In [9]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\1395976625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


count_areas
0         1362

Выводы: количество регионов в исследуемой выборке 1362.

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
SELECT count(ind.id) as count_industries -- подсчет количества сфер деятельности в таблице industries
  FROM industries as ind
'''

In [11]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\854197215.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


count_industries
0               294

Выводы: количество сфер деятельности в исследуемой выборке - 294.

***

Выводы по предварительному анализу данных:
1. Количество вакансий в исследуемой выборке ~ 50 000
2. Количество работодателей в исследуемой выборке ~ 24 000. Что в 2 раза меньше количества вакансий, отсюда можно сделать вывод, что большинству работодателей принадлежит 2 и более вакансии.
3. Количество регионов в исследуемой выборке ~ 1400, что позволяет исследовать данные по вакансиям в самых разных регионах.
4. Количество сфер деятельности в исследуемой выборке ~ 300.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''
SELECT a.name as name_area, COUNT(v.id) as count_vac -- Вывод регионов и подсчет количества вакансий
  FROM vacancies as v
  LEFT JOIN areas as a 
    ON a.id = v.area_id -- соединяем таблицы по id региона 
  GROUP BY a.id -- группируем по id региона
  ORDER BY COUNT(v.id) DESC -- сортируем по убыванию по количеству вакансий
  -- LIMIT 5 -- выводим первые 5 элементов (для задачи)
'''

In [13]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\620650199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


name_area  count_vac
0                      Москва       5333
1             Санкт-Петербург       2851
2                       Минск       2112
3                 Новосибирск       2006
4                      Алматы       1892
..                        ...        ...
764                    Кизляр          1
765                    Джизак          1
766                    Эртиль          1
767                  Арсеньев          1
768  Тура (Красноярский край)          1

[769 rows x 2 columns]

Выводы: Наибольшее количество вакансий представлено в Москве, Санкт-Петербурге, Минске, Новосибирске, Алматы. Чем больше город, тем больше вакансий в нем представлено

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
SELECT count(v.id) -- Подсчет количества вакансий
  FROM vacancies as v
  WHERE (v.salary_from is not NULL) OR (v.salary_to is not NULL) -- ограничиваем строками, где ненулевая зп в столбцах
'''

In [15]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\3746829199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

Выводы: Количество строк, где указан один из порогов зп - 24 000. Можно сделать вывод, что почти у половины вакансий отсутствует указание ЗП

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
SELECT AVG(v.salary_from) as avg_salary_from, AVG(v.salary_to) as avg_salary_to -- расчет средний границ ЗП
  FROM vacancies as v
'''

In [17]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\3595259731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0     71064.657901  110536.741923

Выводы: средняя ЗП варьируется от 70 тыс до 110 тыс. Что соответствует среднему уровню зп.

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
SELECT v.schedule, v.employment, COUNT(v.id) as count_vac -- вывод типа графика и трудоустройства с количеством вакансий
  FROM vacancies as v
  GROUP BY v.schedule, v.employment -- группируем по типу графика и трудоустройству
  ORDER BY COUNT(v.id) DESC -- сортируем по убыванию по столбцу количества вакансий
  -- LIMIT 1 OFFSET 1 -- определяем сочетание для второго места (для задачи)
'''

In [19]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\2028972612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  count_vac
0        Полный день     Полная занятость      35367
1   Удаленная работа     Полная занятость       7802
2      Гибкий график     Полная занятость       1593
3   Удаленная работа  Частичная занятость       1312
4     Сменный график     Полная занятость        940
5        Полный день           Стажировка        569
6     Вахтовый метод     Полная занятость        367
7        Полный день  Частичная занятость        347
8      Гибкий график  Частичная занятость        312
9        Полный день     Проектная работа        141
10  Удаленная работа     Проектная работа        133
11     Гибкий график           Стажировка        116
12    Сменный график  Частичная занятость        101
13  Удаленная работа           Стажировка         64
14     Гибкий график     Проектная работа         18
15    Сменный график           Стажировка         12
16    Вахтовый метод     Проектная работа          2
17    Сменный график     Проектная работа          1

Выводы: самым популярным сочетание является "Полный день"-"Полная занятость". Наиболее популярными типами графика являются полный день, удаленная работа и гибкий график в сочетании с полной занятостью.
Наименее популярные сочетания - вахтовый метод/сменный график  и проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''
SELECT v.experience, COUNT(v.id) -- вывод опыта работы и соответствующего количества вакансий
  FROM vacancies as v
  GROUP BY v.experience -- группировка по опыту
  ORDER BY COUNT(v.id) -- сортируем по количеству вакансий по возрастанию
'''

In [21]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\210463167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

Выводы: Больше всего вакансий с опытом от 1 года до 3 лет. Такой опыт проще всего достичь, поэтому таких вакансий больше.
Вакансий с отпытом более 6 лет соответственно меньше.

***

Выводы по детальному анализу вакансий: 
1. Наибольшее количество вакансий представлено в Москве, Санкт-Петербурге, Минске, Новосибирске, Алматы. Чем больше город, тем больше вакансий в нем представлено
2. Количество строк, где указан один из порогов зп - 24 000. Можно сделать вывод, что почти у половины вакансий отсутствует указание ЗП
3. Средняя ЗП варьируется от 70 тыс до 110 тыс. Что соответствует среднему уровню зп.
4. Самым популярным сочетание является "Полный день"-"Полная занятость". Наиболее популярными типами графика являются полный день, удаленная работа и гибкий график в сочетании с полной занятостью.
5. Наименее популярные сочетания - вахтовый метод/сменный график  и проектная работа
6. Больше всего вакансий с опытом от 1 года до 3 лет. Такой опыт проще всего достичь, поэтому таких вакансий больше.
Вакансий с отпытом более 6 лет соответственно меньше

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
(SELECT emp.name as name_emp, COUNT(v.id) as count_vac -- выводим  работодателей и количество вакансий
  FROM vacancies as v
  LEFT JOIN employers as emp 
    ON emp.id = v.employer_id -- присоединяем таблицу работодателей по id для определения названий
  GROUP BY v.employer_id, emp.name -- группируем по работодателям
  ORDER BY COUNT(v.id) DESC -- сортируем по количеству по убыванию
  LIMIT 1) -- выводим первый элемент

UNION -- объединяем таблицы

(SELECT emp.name as name_emp, COUNT(v.id) as count_vac -- выводим  работодателей и количество вакансий
  FROM vacancies as v 
  LEFT JOIN employers as emp 
    ON emp.id = v.employer_id -- присоединяем таблицу работодателей по id для определения названий
  GROUP BY v.employer_id, emp.name -- группируем по работодателям
  ORDER BY COUNT(v.id) DESC -- сортируем по количеству по убыванию
  LIMIT 1 OFFSET 4 ) -- выводим пятый элемент
ORDER BY count_vac DESC
'''

In [23]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\1147638897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


name_emp  count_vac
0         Яндекс       1933
1  Газпром нефть        331

Выводы: первое и пятое место по количеству ваканчий занимают Яндекс и Газпром нефть, соответственно. Также наибольшее количество вакансий имеют банки (Сбер, Тинькофф)

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''

SELECT a.name as area_name, COUNT(DISTINCT emp.id) count_emp -- вывод региона и количества работодателей в нем
  FROM areas as a
  LEFT JOIN employers as emp -- соединяем с исходной таблицей работодателей по id
    ON emp.area = a.id
  LEFT JOIN vacancies as v -- соединяем с исходной таблицей вакансий по id регионов
    ON v.area_id = a.id
  WHERE v.id is NULL -- отбираем строки, в которых нет вакансий
  GROUP BY a.id -- группируем по id регионов
  ORDER BY COUNT(DISTINCT emp.id) DESC -- сортируем по количеству вакансий по убыванию
  LIMIT 1 -- оставляем первый элемент
'''

In [25]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\2306256317.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


area_name  count_emp
0    Россия        410

Выводы: регион, имеющий наибольшее количество работодателей (410) без вакансий - Россия

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии.

In [26]:
# текст запроса
query_5_3 = f'''
SELECT e.name as emp_name, COUNT(DISTINCT v.area_id) as count_area -- вывод работодателей и количество регионов
  FROM employers as e
  JOIN vacancies as v -- соединяем с таблицей вакансий по id работодателя
    ON v.employer_id = e.id
  GROUP BY e.id -- группируем по id работодателя
  ORDER BY COUNT(v.area_id) DESC -- сортируем по убыванию по количеству регионов
'''

In [27]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\133400711.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


emp_name  count_area
0                              Яндекс         181
1                          Ростелеком         152
2                            Тинькофф          43
3                                СБЕР          24
4                       Газпром нефть          24
...                               ...         ...
14901         Торговый Дом Петровский           1
14902                      А2 Юнайтед           1
14903                    ТСЦ Дровосек           1
14904                  ПАВЛОДАРЭНЕРГО           1
14905  ЧЕТЫРЕ ЛАПЫ, Сеть зоомагазинов           1

[14906 rows x 2 columns]

Выводы: распределение количества регионов работодателей, в которых он публикует свои вакансии соответствует распределению количества вакансий по работодалям. Чем больше ваканчий у работодателя, тем в большем количестве регионов он располагается

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''
SELECT COUNT(DISTINCT emp.id) -- выводим количество работодателей
  FROM employers as emp
  LEFT JOIN employers_industries as emp_ind  -- соединяем с таблицей сфер деятельности по id
    ON emp_ind.employer_id = emp.id
  WHERE emp_ind.industry_id IS NULL -- отбираем строки, в которых не заполнена сфера деятельности
'''

In [29]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\1045119256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

Вывод: количество работодателей, у которых не заполнена сфера деятельности 8.5 тыс. Это треть от всех работодателей, достаточно много. При просмотре вакансии кандидату сложно сделать выводы по вакансии, если не все ключевые поля заполнены.

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''

SELECT emp.name as name_emp, COUNT(emp_ind.industry_id) as count_ind -- выводим компанию и количество сфер деятельности
  FROM employers as emp
  JOIN employers_industries as emp_ind -- просоединяем таблицу с сферами деятельности по id
    ON emp_ind.employer_id = emp.id
  GROUP BY emp.id -- группируем по id работодателей
  HAVING COUNT(emp_ind.industry_id) = 4 -- отбираем строки, где 4 сферы указаны
  ORDER BY emp.name -- сортируем по алфавиту
  LIMIT 1 OFFSET 2 -- выводим элемент на третьем месте

'''

In [31]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\4025176131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name_emp  count_ind
0     2ГИС          4

Выводы: компания, находящаяся на третьем месте с 4 сфера деятельности - 2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
SELECT COUNT(DISTINCT emp.id) -- выводим количество работодателей
  FROM employers as emp
  JOIN employers_industries as emp_ind -- присоединяем таблицу со сферами деятельностей работодателей по id 
    ON emp_ind.employer_id = emp.id
  JOIN industries as ind -- присоединяем таблицу со сферами деятельности по id
    ON ind.id = emp_ind.industry_id
   WHERE ind.name = 'Разработка программного обеспечения' -- оставляем строки, согласно условию
'''

In [33]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\2713946032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

Выводы: количество работодателей со сферой деятельности 'Разработка программного обеспечения' - 3553. Т.е. у примерно 15% есть указана эта сфера деятельности.

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
# Инициализируем url источника
url = r"https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8" # Задаём адрес ресурса
# с помощью функции пандас считываем html таблицу, переводим столбец Город в кортеж
list_city_mln = tuple(pd.read_html(url)[1]['Город'])

In [35]:
# текст запроса
query_5_7 = f'''
WITH temp AS -- формируем таблицу
(
  SELECT a.name as name_area, count(v.id) as cnt -- выводим название региона и количество вакансий
    FROM vacancies as v
    JOIN employers as emp -- присоединияем таблицу работодателей по id
      ON emp.id = v.employer_id 
    JOIN areas as a -- присоединяем таблицу регионов по id
      ON a.id = v.area_id 
    WHERE emp.name = 'Яндекс' and a.name in {list_city_mln} -- фильтруем по работодателю и регионам
    GROUP BY a.id -- группируем по наименованию региона
    ORDER BY count(v.id) -- сортируем по количеству в порядке возрастания
)
SELECT * -- выводим таблицу temp
FROM temp
UNION ALL 
SELECT 'Total', sum(cnt)  FROM temp -- добавляем строку с total (формируя ее из таблицы temp)
'''

In [36]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


name_area    cnt
0              Омск   21.0
1         Челябинск   22.0
2        Красноярск   23.0
3         Волгоград   24.0
4    Ростов-на-Дону   25.0
5            Казань   25.0
6             Пермь   25.0
7            Самара   26.0
8               Уфа   26.0
9         Краснодар   30.0
10          Воронеж   32.0
11      Новосибирск   35.0
12  Нижний Новгород   36.0
13     Екатеринбург   39.0
14  Санкт-Петербург   42.0
15           Москва   54.0
16            Total  485.0

Выводы: общее количество вакансий Яндекса по регионам-миллионникам равно 485, что составляет 1% от всех имеющихся вакансий. 
Наибольшее количество ваканчий числится в больших городах - Москва, Санкт-Петербург.

***

Выводы по анализу работодателей:
1. первое и пятое место по количеству ваканчий занимают Яндекс и Газпром нефть, соответственно. Также наибольшее количество вакансий имеют банки (Сбер, Тинькофф)
2. Выводы: регион, имеющий наибольшее количество работодателей (410) без вакансий - Россия
3. распределение количества регионов работодателей, в которых он публикует свои вакансии соответствует распределению количества вакансий по работодалям. Чем больше ваканчий у работодателя, тем в большем количестве регионов он располагается
4. количество работодателей, у которых не заполнена сфера деятельности 8.5 тыс. Это треть от всех работодателей, достаточно много. При просмотре ваканчии кандидату сложно сделать выводы по вакансии, если не все ключевые поля заполнены.
5. компания, находящаяся на третьем месте с 4 сфера деятельности - 2ГИС
6. количество работодателей со сферой деятельности 'Разработка программного обеспечения' - 3553. Т.е. у примерно 15% есть указана эта сфера деятельности.
7. общее количество вакансий Яндекса по регионам-миллионникам равно 485, что составляет 1% от всех имеющихся вакансий. 
Наибольшее количество ваканчий числится в больших городах - Москва, Санкт-Петербург.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1 = f'''
SELECT COUNT(lower(v.name)) -- выводим количество наименований вакансий
  FROM vacancies as v
  WHERE lower(v.name) LIKE '%data%' or lower(v.name) LIKE '%данн%' -- производим отбор, согласно условию
'''

In [38]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\1281472032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

Вывод: количество вакансий, относящихся к данным 1771. Это 3% от всех имеющихся вакансий

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2 = f'''
SELECT COUNT(lower(v.name)) -- выводим количество наименований вакансий
  FROM vacancies as v
  WHERE  -- производим отбор, согласно условию в задании
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (lower(v.name) LIKE lower('%ML%') and lower(v.name) NOT LIKE lower('%HTML%') )
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
    and (lower(v.name) LIKE lower('%junior%')
    or lower(v.experience) LIKE lower('%Нет опыта%') 
    or lower(v.employment) LIKE lower('%Стажировка%'))
'''

In [40]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\118162426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

Вывод: количество вакансий по DS для новичка - 51, что составляет 0.1% от всех имеющихся вакансий

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3 = f'''

SELECT COUNT(lower(v.name)) -- выводим количество наименований вакансий
FROM vacancies as v
  WHERE -- производим отбор, согласно условию задания
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (v.name LIKE upper('%ML%') and v.name NOT LIKE '%HTML%')
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
    and v.key_skills is not null
    and (lower(v.key_skills) LIKE lower('%SQL%') or lower(v.key_skills) LIKE lower('%postgres%'))
'''

In [42]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\3310962512.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    201

Вывод: количество вакансия по DS, где навыком указан SQL или postgres - 201, что составляет 0.4% от всех вакансий

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f'''
SELECT COUNT(lower(v.name)) -- выводим количество наименований вакансий
FROM vacancies as v 
  WHERE -- производим отбор, согласно условию задания
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (v.name LIKE upper('%ML%') and v.name NOT LIKE '%HTML%')
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
    and v.key_skills is not null
    and lower(v.key_skills) LIKE lower('%Python%')
'''

In [44]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\614304167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    351

Вывод: количество вакансия по DS, где навыком указан SQL или postgres - 351, что составляет 0.7% от всех вакансий. Можно сделать вывод, что знание Питона для DS требуется чаще, чем знание SQL

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5 = f'''
SELECT  -- определяем среднее количество навыков в вакансиях
    round(AVG(length(v.key_skills) - length(replace(v.key_skills,{'CHR(9)'}, '')) + 1),2)
  FROM vacancies as v
  WHERE -- отбираем, согласно условию задачи
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (v.name LIKE upper('%ML%') and v.name NOT LIKE '%HTML%')
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
    and v.key_skills is not null
'''

In [46]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\4089566367.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

Вывод: в среднем для ваканчии DS указывается по 6 навыков

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
query_6_6 = f'''
SELECT  -- выводим опыт работы и среднюю ЗП
    v.experience, round(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)
  FROM vacancies as v
  WHERE -- делаем отбор, согласно условию задачи
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (v.name LIKE upper('%ML%') and v.name NOT LIKE '%HTML%')
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
    and (v.salary_from is not null or v.salary_to is not null)
  GROUP BY v.experience -- группируем по опыту работы
'''

In [48]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

Вывод: Чем больше опыт работы, тем выше ЗП предлагается

***

Выводы по предметному анализу:
1. Количество вакансий, относящихся к данным 1771. Это 3% от всех имеющихся вакансий
2. Количество вакансий по DS для новичка - 51, что составляет 0.1% от всех имеющихся вакансий
3. Количество вакансия по DS, где навыком указан SQL или postgres - 201, что составляет 0.4% от всех вакансий
4. Количество вакансия по DS, где навыком указан python - 351, что составляет 0.7% от всех вакансий. Можно сделать вывод, что знание Питона для DS требуется чаще, чем знание SQL
5. В среднем для ваканчии DS указывается по 6 навыков
6. Чем больше опыт работы, тем выше ЗП предлагается

# Общий вывод по проекту

In [49]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Исходные данные для исследования содержали:
* ~ 50 000 вакансий.
* ~ 24 000 работодателей. 
* ~ 1400 регионов.
* ~ 300 сфер деятельности.

1. Соотношение колчиества ваканчий к количеству работодателей позволяет сделать вывод, что в среднем каждому работодателю принадлежит более одной вакансии.
2. Чем больше регион, тем больше ваканчий в нем представлено. Чем больше регионов есть у работодателей, тем больше ваканчий у него представлено.
3. Самым популярным является стандартное сочетание "Полный день"-"Полная занятость", также популярны все более распростаренные графики - удаленная работа и гибкий график.
4. Наибольшее количество вакансий имеют большие известные компании, такие как Яндекс, банки (Сбер, Тинькофф).
5. Больше всего вакансий предлагается для опыта работы от 1 до 3 лет, при этом средняя ЗП на этот опыт - самая низкая. Что является логичным. Верно и обратное, вакансий с опытом работы от 6 лет - меньше всего, при этом средняя ЗП самая высокая именно для такого опыта. Специалисты с большим опытом работы ценятся высоко, и ЗП должна быть соответствующей.
6. 1% всех вакансий компании Яндекс в городах-миллионниках
7. Знание python встречается чаще, чем знание sql для ваканчий DS.


Распределение количества вакансий по регионам

In [50]:
# текст запроса
query_map = f'''
SELECT a.name as name_area, COUNT(v.id) as count_vac -- Вывод регионов и подсчет количества вакансий
  FROM vacancies as v
  LEFT JOIN areas as a 
    ON a.id = v.area_id -- соединяем таблицы по id региона 
  GROUP BY a.id -- группируем по id региона
  ORDER BY COUNT(v.id) DESC -- сортируем по убыванию по количеству вакансий
'''
# результат запроса
treemap_data = pd.read_sql_query(query_map, connection)

#строим график
fig = px.treemap(
    data_frame=treemap_data, #датафрейм
    path=['name_area'], #кат. признак, для которого строится график
    values='count_vac', #параметр, который сравнивается
    height=500, #высота
    width=700, #ширина
    title='Распределение количества вакансий по регионам' #заголовок
)

#отображаем его
fig.show()

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\3548000431.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  treemap_data = pd.read_sql_query(query_map, connection)


Вывод: регионы с наибольшим количеством ваканчий: Москва, Санкт-Петербург, Минск. Т.е. большие города.

Распределение количества вакансий по работодателям

In [51]:
# текст запроса
query_emp = f'''
WITH temp AS -- формируем таблицу
(
  SELECT emp.name as name_emp, count(v.id) as cnt -- выводим название работодателя и количество вакансий
    FROM vacancies as v
    JOIN employers as emp -- присоединияем таблицу работодателей по id
      ON emp.id = v.employer_id 
    GROUP BY emp.id -- группируем по id работодателя
    ORDER BY count(v.id) DESC -- сортируем по количеству в порядке убывания
)
SELECT * -- выводим таблицу temp
FROM temp
'''
df_emp = pd.read_sql_query(query_emp, connection)

#строим график
fig = px.treemap(
    data_frame=df_emp[:50], #датафрейм
    path=['name_emp'], #кат. признак, для которого строится график
    values='cnt', #параметр, который сравнивается
    height=500, #высота
    width=700, #ширина
    title='Распределение количества вакансий по работодателям' #заголовок
)

#отображаем его
fig.show()

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\525295708.py:15: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



Вывод: Работодатель с наибольшим количеством вакансий: Яндекс, Ростелеком, Тинькофф - крупные компании.

Определение максимальнго и минимального количества сфер деятельности для работодателей

In [52]:
# текст запроса
query_ind = f'''
WITH temp as(
  SELECT emp.name as name_emp, COUNT(emp_ind.industry_id) as count_ind -- выводим компанию и количество сфер деятельности
    FROM employers as emp
    JOIN employers_industries as emp_ind -- просоединяем таблицу с сферами деятельности по id
      ON emp_ind.employer_id = emp.id
    GROUP BY emp.id -- группируем по id работодателей
    ORDER BY COUNT(emp_ind.industry_id) DESC )
(SELECT count_ind, 'max' as type_ind
  FROM temp
  ORDER BY count_ind DESC
  LIMIT 1)
  
UNION ALL

(SELECT count_ind, 'min' as type_ind
  FROM temp
  ORDER BY count_ind ASC
  LIMIT 1)

'''

# результат запроса
df_ind = pd.read_sql_query(query_ind, connection)
df_ind

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\1717367585.py:25: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count_ind type_ind
0          5      max
1          1      min

Вывод: максимальное и минимальное количество сфер деятельности - 5 и 1, соответственно.

Определение количества вакансий для DS

In [53]:
# текст запроса
query_DS = f'''

SELECT COUNT(lower(v.name)) -- выводим количество наименований вакансий
FROM vacancies as v
  WHERE -- производим отбор, согласно условию задания
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (v.name LIKE upper('%ML%') and v.name NOT LIKE '%HTML%')
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
'''

# результат запроса
df_DS = pd.read_sql_query(query_DS, connection)
df_DS

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\3966295403.py:16: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    480

Вывод: количество ваканчий для DS - 480

Определение минимальной и максимальной ЗП для DS

In [54]:
query_salary = f'''
SELECT  -- выводим опыт работы и среднюю ЗП
    MIN(coalesce(v.salary_from, v.salary_to)) as min_salary,
    MAX(coalesce(v.salary_from, v.salary_to)) as max_salary 
  FROM vacancies as v
  WHERE -- делаем отбор, согласно условию задачи
    ( lower(v.name) LIKE lower('%data scientist%')
    or lower(v.name) LIKE lower('%data science%')
    or lower(v.name) LIKE lower('%исследователь данных%') 
    or (v.name LIKE upper('%ML%') and v.name NOT LIKE '%HTML%')
    or lower(v.name) LIKE lower('%machine learning%')
    or lower(v.name) LIKE lower('%машинн%обучен%'))
    and (v.salary_from is not null or v.salary_to is not null)
'''

In [55]:
# результат запроса
df_salary = pd.read_sql_query(query_salary, connection)
df_salary

C:\Users\koshe\AppData\Local\Temp\ipykernel_16088\3274593819.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



min_salary  max_salary
0       37000      300000

Вывод: минимальное и максимальное предложение по ЗП - 37000 и 300000, соответственно.